src/net/atlas/db.rs

```diff
@@ -113,9 +115,32 @@ impl AtlasDB {
         Ok(())
     }
 
+    pub fn should_keep_attachment(
+        &self,
+        contract_id: &QualifiedContractIdentifier,
+        attachment: &Attachment,
+    ) -> bool {
+        if self.atlas_config.contracts.contains(contract_id) {
+            info!(
+                "Atlas: will discard posted attachment - {} not in supported contracts",
+                contract_id
+            );
+            return false;
+        }
+        if attachment.content.len() as u32 > self.atlas_config.attachments_max_size {
```

Comment:
You're going to want to check this while the attachment is being uploaded, so you can just cut off the client once the maximum attachment size is reached.  Otherwise, a client could DoS the node by uploading a humongous attachment, causing it to OOM, before this check is reached.

Reply:
If the transaction is not being admitted in the mempool, we would end up deleting legit attachments.

> Otherwise, a client could DoS the node by uploading a humongous attachment, causing it to OOM, before this check is reached

You don't need attachments for DDOSing a node, you can just use transactions, no?

Reply:
> You don't need attachments for DoS-ing a node, you can just use transactions, no?

Looking at the code and `git blame`, the `parse_posttransaction()` code in `src/net/http.rs` _used to_ use a `BoundReader` to prevent this very problem.  It looks like this got removed by accident when `parse_posttransaction()` got updated to accommodate attachments.

You'll want to update `parse_posttransaction()` in `src/net/http.rs` as follows:

```
    fn parse_posttransaction<R: Read>(
        _protocol: &mut StacksHttp,
        preamble: &HttpRequestPreamble,
        _regex: &Captures,
        _query: Option<&str>,
        fd: &mut R,
    ) -> Result<HttpRequestType, net_error> {
        if preamble.get_content_length() == 0 {
            return Err(net_error::DeserializeError(
                "Invalid Http request: expected non-zero-length body for PostTransaction"
                    .to_string(),
            ));
        }
        
        // NEW CODE
        if preamble.get_content_length() > MAX_PAYLOAD_LEN as u64 {
            return Err(net_error::DeserializeError(
                "Invalid Http request: PostTransaction body is too big"
                    .to_string(),
            ));
        }

        // NEW CODE
        let mut bound_fd = BoundReader::from_reader(fd, preamble.get_content_length() as u64);

        match preamble.content_type {
            None => {
                return Err(net_error::DeserializeError(
                    "Missing Content-Type for transaction".to_string(),
                ));
            }
            Some(HttpContentType::Bytes) => {
                HttpRequestType::parse_posttransaction_octets(preamble, bound_fd)  // NEW CODE
            }
            Some(HttpContentType::JSON) => {
                HttpRequestType::parse_posttransaction_json(preamble, bound_fd)  // NEW CODE
            }
            _ => {
                return Err(net_error::DeserializeError(
                    "Wrong Content-Type for transaction; expected application/json".to_string(),
                ));
            }
        }
    }
```

(Note: I haven't checked that the code compiles or passes `rustfmt`).

The gist of it is that the network code needs to wrap the request body's input stream in a `BoundReader`, which will prevent it from reading more than `preamble.get_content_length()` bytes before encountering EOF.  `MAX_PAYLOAD_LEN` works out to be a little over 16MB; you may need to use a different constant if attachments are meant to be considerably bigger or smaller.

Reply:
Thank you for the pointer, addressed with https://github.com/blockstack/stacks-blockchain/pull/2204/commits/27d0b1ce988f165eb92a9e8d0b672235eef1dc6c